### PR TITLE
Add docker files

### PR DESCRIPTION
**Pull request by haitaohuang**
Add docker files and scripts to demonstrate how to:
1. build psw, sdk from source, then install in docker and run sample and aesm in containers.
2. install psw, sdk from releases hosted on 01.org, then run sample and aesm in containers.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +28,19 @@ Documentation
 - [Intel(R) SGX for Linux\* OS](https://01.org/intel-softwareguard-extensions) project home page on [01.org](https://01.org)
 - [Intel(R) SGX Programming Reference](https://software.intel.com/sites/default/files/managed/7c/f1/332831-sdm-vol-3d.pdf)
 
+Quick Start
+-----------------------------------------
+### Use docker and docker composer
```
Comment (rodolfoams) [inline README.md]: ### Use Docker and Docker Compose

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+
+FROM ubuntu:18.04 as builder
+
+RUN apt-get update --fix-missing
+RUN apt-get install -y build-essential ocaml ocamlbuild automake autoconf libtool wget python libssl-dev 
+RUN apt-get install -y git libcurl4-openssl-dev protobuf-compiler libprotobuf-dev debhelper cmake lsb-release
```
Comment (rodolfoams) [inline docker/Dockerfile]: I was only able to build this image after adding `curl` as a dependency package here.

File: docker/README.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
+# Build and run with Docker
+
+Files in this directory demostrate how to build SGX SDK and PSW, then install and run SGX applications in docker containers.
+
+## Quick Start
+
+###  Prerequisites
+1. Install [docker and composer](https://docs.docker.com/) and configure them properly following respective installation guide.
```
Comment (rodolfoams) [inline docker/README.md]: Docker and Compose

File: docker/README.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+2. Install [SGX out-of-tree driver](https://github.com/intel/linux-sgx-driver). **Note**: See below to run with DCAP driver or SGX capable kernel.
+
+### Run with docker composer
```
Comment (rodolfoams) [inline docker/README.md]: Docker Compose

File: docker/README.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+This will start aesm and sample from one terminal using docker composer.
```
Comment (rodolfoams) [inline docker/README.md]: docker-compose

File: docker/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+In second terminal, 
+```
+$ ./build_and_run_sample_docker.sh
+```
+## Helper scripts
+
+These shell scripts demostrate steps to build SDK and PSW, theh start containers.
+- [build_and_run_aesm_docker.sh](../docker/build_and_run_aesm_docker.sh) - build and run AESM in docker, using bin installer
+- [build_and_run_sample_docker.sh](../docker/build_and_run_sample_docker.sh) - build and run SampleEnclave example in docker, using bin installer
+- [build_compose_run.sh](../docker/build_compose_run.sh) - use docker composer to bring up separate containers for AESM and SampleEnclave example.
```
Comment (rodolfoams) [inline docker/README.md]: docker-compose

File: docker/README.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+The Dockerfile is a multi-stage docker file that specifies 3 image build targets:
+1. builder: builds psw and sdk bin installers from source, with necessary prebuilt AEs and libs downloaded from 01.org.
+2. aesm: takes psw installer from builder, install and run the aesm deamon.
+3. sample: installs sdk installer from builder, build and run the SampleEnclave app
+
+build_and_run_aesm_docker.sh shows how to build and run the aesm image. This will start an aesm service listening to named socket, mounted to /var/run/aesmd in the container from host /tmp/aesmd
+
+build_and_run_sample_docker.sh shows how to build and run the SampleEnclave app inside container with locally built sgx_sample image.
+
+## Docker composer files
```
Comment (rodolfoams) [inline docker/README.md]: Docker Compose

File: docker/README.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+docker-compose.yml can be used with docker composer to start both aesm and sample app in separate containers.
```
Comment (rodolfoams) [inline docker/README.md]: docker-compose

File: linux/installer/docker/README.md [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# Deploy SGX enclaves in containers
+
+Files in this directory demostrate how to deploy SGX enclave applications in docker containers.
+
+## Quick Start
+
+###  Prerequisites
+1. Install [docker and composer](https://docs.docker.com/) and configure them properly following respective installation guide.
```
Comment (rodolfoams) [inline linux/installer/docker/README.md]: Docker and Compose

File: linux/installer/docker/README.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+2. Install [SGX out-of-tree driver](https://github.com/intel/linux-sgx-driver). **Note**: See below to run with DCAP driver or SGX capable kernel.
+
+### Run with docker composer
```
Comment (rodolfoams) [inline linux/installer/docker/README.md]: docker-compose

File: linux/installer/docker/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+This will start aesm and sample from one terminal using docker composer.
```
Comment (rodolfoams) [inline linux/installer/docker/README.md]: docker-compose

File: docker/Dockerfile (hunk 2 above)
Comment (haitaohuang) [reply docker/Dockerfile]: hmm, can you tell which step did it fail or the error message?

**Comment by haitaohuang** [comment]
will correct those names. Thanks

File: docker/Dockerfile (hunk 2 above)
Comment (rodolfoams) [reply docker/Dockerfile]: This is the step and error message obtained:

```bash
Step 12/21 : RUN make psw_install_pkg
 ---> Running in 44272a018f53
fatal: not a git repository (or any of the parent directories): .git
curl --output dcap_source.tar.gz -L --tlsv1 https://github.com/intel/SGXDataCenterAttestationPrimitives/archive/DCAP_1.4.tar.gz
make: curl: Command not found
Makefile:62: recipe for target 'dcap_source' failed
make: *** [dcap_source] Error 127
The command '/bin/sh -c make psw_install_pkg' returned a non-zero code: 2
```

Steps to reproduce error:

```bash
git clone https://github.com/haitaohuang/linux-sgx.git -b docker_pr3
cd linux-sgx/docker
docker build --target aesm -t sgx_aesm -f ./Dockerfile ../
```

Comment (haitaohuang) [reply docker/Dockerfile]: Ah, thanks
I used git clone --recurse-submodules.

**Comment by haitaohuang** [comment]
note to reviewers: I'll keep adding patches for fixes/updates to this branch for now. Will squash before merge

**Comment by haitaohuang** [comment]
Squashed and removed reproducible-build

**Pull request closed without merging**